proposals/0000-record-with-syntax.rst

```diff
@@ -105,49 +105,72 @@ The feature is well precedented by layout syntax in other Haskell constructs.
 
 Proposed Change Specification
 -----------------------------
+
+Informal Specification
+~~~~~~~~~~~~~~~~~~~~~~
 We propose a language extension called ``RecordWith`` offering layout syntax for records.
-With the extension enabled, there is an additional keyword ``with`` which induces layout syntax for record fields.
+With the extension enabled, there is an additional keyword ``with`` inducing layout syntax for record fields.
+
+The ``with`` keyword is a so-called *layout herald*: it enables the layout mechanism in the same way as ``where``, ``let``, ``do`` and ``of``.
+In the formal grammar, ``with`` should be followed by curly braces ``{ }`` and semicolons ``;`` delimiting fields.
+In the source program, the user can instead use indentation and newlines as delimiters, and the layout mechanism will automatically insert the missing braces and semicolons.
+A general description of the layout mechanism is available `here <https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-210002.7>`_.
+
+To achieve a uniform record syntax, we propose to allow semicolons as field delimiters even without the ``with`` keyword.
+(However you cannot mix commas and semicolons in a single field list.)
+We allow optional leading and trailing semicolons, as well as extraneous repeated semicolons, for consistency with other layout syntax.
+The more flexible semicolon delimiters could be useful for easier record editing, even without layout syntax.
 
-The structure and formatting options are very similar to monadic ``do`` blocks (and other layout-inducing constructs).
-Specifically, ``with`` can be followed either by indentation and newline delimiters or by curly braces ``{ }`` and semicolon ``;`` delimiters.
-For example, the following 3 declarations are all valid, equivalent ways to define a record type ``Rec``:
+The syntax proposal is easiest to understand through examples.
+Firstly, we can declare a record ``Rec`` using the layout herald ``with`` and newline delimiters.
 ::
 
   data Rec = Rec with
     field1 :: T1
     field2 :: T2
     field3 :: T3
 
+The layout mechanism inserts the necessary punctuation to get the following equivalent declaration.
+::
+
   data Rec = Rec with
     { field1 :: T1
     ; field2 :: T2
     ; field3 :: T3 }
-  
+
+Alternatively, you can write a hybrid version all on one line as follows.
+::
+
   data Rec = Rec with field1 :: T1; field2 :: T2; field3 :: T3
 
-These are all equivalent to the traditional syntax
+Here the layout mechanism inserts the curly braces for us.
+
+We can also use semicolon delimiters without the ``with`` layout herald:
 ::
 
-  data Rec = Rec { field1 :: T1, field2 :: T2, field3 :: T3 }
+  data Rec = Rec { field1 :: T1; field2 :: T2; field3 :: T3 }
```

Comment:
This option, using semicolon delimiters without the `with` herald seems inconsistent with the other layout-aware keywords in the language. Perhaps we should limit the use of semicolon delimiters to the new `with` syntax?

Reply:
Could go either way, that does sound reasonable to me. I assume that most people don’t think about `;` when they see layout code, so emphasizing the `;` here too much does not seem to be useful.